### DISCUSSION TITLE
Necessity of `async` test function with concurrent mode

### DISCUSSION THREAD

**Discussion by xsjcTony** [vitest-dev/vitest#6072] [Q&A] [answered]
The documentation said

> It receives the test name, an async function with the tests to collect, and an optional timeout (in milliseconds).

However, it's not strictly enforced on the type level (which means you can still use the normal non-async function together with `.concurrent`).

Wondering whether there is any harm to use non-async function and `.concurrent` together? Will Vitest wrap the test function into async when actually collecting / running the test?

**Comment by sheremet-va** [comment]
It will wrap it in an async function but there is no benefit in running two sync functions concurrently because JS is single-threaded - it will still run them one after another.

**Reply by xsjcTony** [reply]
So let's take the example below, which is the example from the documentation.
I intentionally removed all `async` marks for test functions (assuming no `await` needed in the test function), will it behave differently?
```ts
import { describe, it } from 'vitest'

// All tests within this suite will be run in parallel
describe.concurrent('suite', () => {
  it('concurrent test 1', ({ expect }) => { /* ... */ })
  it('concurrent test 2', ({ expect }) => { /* ... */ })
  it.concurrent('concurrent test 3', ({ expect }) => { /* ... */ })
})
```
Originally:
```ts
import { describe, it } from 'vitest'

// All tests within this suite will be run in parallel
describe.concurrent('suite', () => {
  it('concurrent test 1', async ({ expect }) => { /* ... */ })
  it('concurrent test 2', async ({ expect }) => { /* ... */ })
  it.concurrent('concurrent test 3', async ({ expect }) => { /* ... */ })
})
```

**Reply by sheremet-va** [reply]
They will all run one after another just because of how JS works.

**Reply by xsjcTony** [reply]
Yeah it makes sense since JS just works this way, but if that's the case, what's the term `concurrent` exactly mean here?

**Reply by xsjcTony** [reply]
If they will eventually run one after another (not in parallel?), which means using `.concurrent` makes no improvement?
```ts
import { describe, it } from 'vitest'

// The two tests marked with concurrent will be run in parallel
describe('suite', () => {
  it('serial test', async () => { /* ... */ })
  it.concurrent('concurrent test 1', async ({ expect }) => { /* ... */ })
  it.concurrent('concurrent test 2', async ({ expect }) => { /* ... */ })
})
```

**Reply by sheremet-va** [reply]
It means what it says. All functions are called at the same time. If all functions have a `fetch` call, for example, Vitest will not wait until it's finished individually. It basically calls `Promise.all([fn1(), fn2(), fn3()])`.

**Reply by xsjcTony** [reply, marked as answer]
Right. Can you please🙏 let me know whether I'm correct about my understanding:

- No matter the test function is `async` or not, Vitest just simply making them `async` and push them to the micro-task queue as long as it's marked with `.concurrent`.
- But because of how JS event loop works, even if semantically they are running in parallel, but in reality they are still executing one by one in the micro-task queue.
- For those non-concurrent test cases, they are executing synchronously straight away (macro-task queue? without pushing to micro-task queue), which means async tasks like `fetch` will be awaited until running next serial test
- Concurrent ≠ Multi threading in Vitest's context.
- Concurrent will not technically improve the performance if there's no async actions like `fetch` inside the test function
- But multi-threading like Sharding does 100% improve the performance

**Reply by sheremet-va** [reply]
You are correct

**Reply by xsjcTony** [reply]
Got it. Thanks for your patience!
